libkbfs/mdserver_branch_journal.go

```
@@ -0,0 +1,159 @@
+	o, err := revisionToOrdinal(r)
+	if err != nil {
+		return err
+	}
+	return j.j.writeEarliestOrdinal(o)
+}
+
+func (j mdServerBranchJournal) readLatestRevision() (
+	MetadataRevision, error) {
+	o, err := j.j.readLatestOrdinal()
+	if os.IsNotExist(err) {
+		return MetadataRevisionUninitialized, nil
+	} else if err != nil {
+		return MetadataRevisionUninitialized, err
+	}
+	return ordinalToRevision(o)
+}
+
+func (j mdServerBranchJournal) writeLatestRevision(r MetadataRevision) error {
+	o, err := revisionToOrdinal(r)
+	if err != nil {
+		return err
+	}
+	return j.j.writeLatestOrdinal(o)
+}
+
+func (j mdServerBranchJournal) readMdID(r MetadataRevision) (MdID, error) {
+	o, err := revisionToOrdinal(r)
+	if err != nil {
+		return MdID{}, err
+	}
+	e, err := j.j.readJournalEntry(o)
+	if err != nil {
+		return MdID{}, err
+	}
+
+	// TODO: Validate MdID?
+	return e.(MdID), nil
+}
+
+// All functions below are public functions.
+
+func (j mdServerBranchJournal) journalLength() (uint64, error) {
+	return j.j.journalLength()
+}
+
+func (j mdServerBranchJournal) getHead() (MdID, error) {
+	latestRevision, err := j.readLatestRevision()
+	if err != nil {
+		return MdID{}, err
+	} else if latestRevision == MetadataRevisionUninitialized {
+		return MdID{}, nil
+	}
+	return j.readMdID(latestRevision)
+}
+
+func (j mdServerBranchJournal) getRange(
+	start, stop MetadataRevision) (MetadataRevision, []MdID, error) {
+	earliestRevision, err := j.readEarliestRevision()
+	if err != nil {
+		return MetadataRevisionUninitialized, nil, err
+	} else if earliestRevision == MetadataRevisionUninitialized {
+		return MetadataRevisionUninitialized, nil, nil
+	}
+
+	latestRevision, err := j.readLatestRevision()
+	if err != nil {
+		return MetadataRevisionUninitialized, nil, err
+	} else if latestRevision == MetadataRevisionUninitialized {
+		return MetadataRevisionUninitialized, nil, nil
+	}
+
+	if start < earliestRevision {
+		start = earliestRevision
+	}
+
+	if stop > latestRevision {
```

It's a little weird that if `stop < earliestRevision`, we'll return a `nil` error, `start == earliestRevision` and an empty range.  Maybe we should return an error if `stop < earliestRevision`?  Does any caller rely on that behavior (say to get the value of `earliestRevision`)?